src/pybind/mgr/orchestrator/_interface.py

```diff
@@ -786,6 +786,7 @@ def __init__(self,
                  osdspec_affinity: Optional[str] = None,
                  last_deployed: Optional[datetime.datetime] = None,
                  events: Optional[List['OrchestratorEvent']] = None,
+                 event_dict: Optional[List[dict]] = None,
```

Comment:
Not sure if this is a good idea.. you are duplicating a _"data source"_ just because you want a different format for the date source.
In any case, I think that the correct way to do this is to keep only one attribute with the source, and to use a new method to return the information in the format you want (dict).
In this case, even I have real doubts about the need of a new_method ... because you have in the DaemonDescription class the method "to_json"... look how easy is to get the event information as a dict:

```
d = DaemonDescription(...)
event_dict = json.load(d.to_json())[events]
```

Reply:
Thanks @jmolmo . I tried this out at first but it seems that `json.load` does not work in this case as we are getting events as string which cannot be converted to json or dict. As we can have only one return type for 'to_json' method i.e. string or dict, so we have two options here: either to have both data sources (string and dict) or we need to change the  return type of events from string to dict. Am not sure what else can be done here. @epuertat @alfonsomthd  can you please give your inputs here.

Reply:
> Thanks @jmolmo . I tried this out at first but it seems that `json.load` does not work in this case as we are getting events as string which cannot be converted to json or dict. [...]

From my understanding it seems you found some misbehavior here:
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/orchestrator/_interface.py#L1247

Can you provide an example so @sebastian-philipp or @jmolmo can take a look also?

Reply:
`2021-04-26T21:04:44.929140Z service:prometheus [INFO] "service was created"`. This is the string that we are getting from this method. I tried 'json.loads' to convert this string to json but as this isn't in 'string-json' format, so this cant be converted to dict.

Reply:
works for me! 

```
mgr $  . .tox/py3/bin/activate
(py3) mgr $  PYTHONPATH=.. UNITTEST=true python
Python 3.9.2 (default, Feb 20 2021, 00:00:00) 
[GCC 10.2.1 20201125 (Red Hat 10.2.1-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tests
>>> from orchestrator import OrchestratorEvent
>>> OrchestratorEvent.from_json('2021-04-26T21:04:44.929140Z service:prometheus [INFO] "service was created"').message
'service was created'
```

Reply:
I think the conflict here is bcz of the need of 2 different return types: for dashboard it requires dict return type where actually to_json return string here
https://github.com/ceph/ceph/blob/9dcf6f7c8cd000ea9c404b5538e1f4d5ade53d02/src/pybind/mgr/dashboard/controllers/service.py#L35
And in order to not change the return type or bringing new data source(another copy of events) , I think it can be resolved by addressing a boolean here (to_json of OrchestratorEvent class)
`def to_json(self, is_dict=False) -> OrderedDict:`
        `if self.events:`
           `if is_dict:`
               `out['events'] = [e.to_dict() for e in self.events]`
            `else:`
                `out['events'] = [e.to_json() for e in self.events]`
And as dashboard needs dict return type we can pass True for is_dict here:
`return [service.to_json(True) for service in orch.services.list(service_name=service_name)]`

to_dict() will look something like this:
`def to_dict(self) -> dict:`
        `created = datetime_to_str(self.created)`
        `return {`
           ` 'created': created,`
            `'subject': self.kind_subject(),`
            `'level': self.level,`
            `'message': self.message`
        `}`
@aaSharma14 what u think will this work?

Reply:
> works for me!
> 
> ```
> mgr $  . .tox/py3/bin/activate
> (py3) mgr $  PYTHONPATH=.. UNITTEST=true python
> Python 3.9.2 (default, Feb 20 2021, 00:00:00) 
> [GCC 10.2.1 20201125 (Red Hat 10.2.1-9)] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import tests
> >>> from orchestrator import OrchestratorEvent
> >>> OrchestratorEvent.from_json('2021-04-26T21:04:44.929140Z service:prometheus [INFO] "service was created"').message
> 'service was created'
> ```

I tried this approach as suggested by you earlier. The problem I faced is when we return the value of `out['events'] = [OrchestratorEvent.from_json(e.to_json()) for e in events]`, it gives OrchestratorEvent object is not JSON Serializable error.

Reply:
> I think the conflict here is bcz of the need of 2 different return types: for dashboard it requires dict return type where actually to_json return string here
> https://github.com/ceph/ceph/blob/9dcf6f7c8cd000ea9c404b5538e1f4d5ade53d02/src/pybind/mgr/dashboard/controllers/service.py#L35
> 
> 
> And in order to not change the return type or bringing new data source(another copy of `events`) , I think it can be resolved by addressing a boolean here (to_json of OrchestratorEvent class)
> `def to_json(self, is_dict=False) -> OrderedDict:`
> `........`
> `if self.events:`
> `if is_dict:`
> `out['events'] = [e.to_dict() for e in self.events]`
> `else:`
> `out['events'] = [e.to_json() for e in self.events]`
> And as dashboard needs dict return type we can pass True for is_dict here:
> `return [service.to_json(True) for service in orch.services.list(service_name=service_name)]`
> to_dict() will look something like this:
> `def to_dict(self) -> dict:`
> `created = datetime_to_str(self.created)`
> `return {`
> ` 'created': created,`
> `'subject': self.kind_subject(),`
> `'level': self.level,`
> `'message': self.message`
> `}`
> @aaSharma14 what u think will this work?

This is working fine for me. Thanks a lot @avanthakkar . @sebastian-philipp @jmolmo Is this approach fine with you?

Reply:
@avanthakkar, the idea of something like `to_json(True)` returning anything other than a JSON string makes me shiver... We shouldn't bury ad-hoc functionality inside the existing methods... It costs nothing to extract the common code and to come up with 2 separate methods (`to_json()` and `to_dict()`), but we should dispel the idea of reusing functions for doing many unrelated things.

Reply:
Ok I see that the original function is already not returning a JSON string... In any case, I don't think we should insist in that error, but rather create a separate `to_dict()` function, that really does what it promises. We can pull the shared code out to another function.

Reply:
> Ok I see that the original function is already not returning a JSON string... In any case, I don't think we should insist in that error, but rather create a separate `to_dict()` function, that really does what it promises. We can pull the shared code out to another function.

Yes I agree. I had 2 approaches one to make duplicate of `to_json` named `to_dict` but just return events as `dict` instead of `string` and the other to support both functionality in just one func- `to_json` and I chose later to not duplicate the functionality just for a having `events` as a different data type, but yeah as you mentioned that's not a good approach here.